src/analyses/relationAnalysis.ml

```diff
@@ -0,0 +1,453 @@
+      Option.default false (ID.to_bool is_eq)
+    | Queries.MayBeEqual (exp1,exp2) ->
+      let exp = (BinOp (Cil.Eq, exp1, exp2, TInt (IInt, []))) in
+      let is_neq = eval_int exp (no_overflow ctx exp) in
+      Option.default true (ID.to_bool is_neq)
+    | Queries.MayBeLess (exp1, exp2) ->
+      let exp = (BinOp (Cil.Lt, exp1, exp2, TInt (IInt, []))) in
+      let is_lt = eval_int exp (no_overflow ctx exp) in
+      Option.default true (ID.to_bool is_lt)
+    | _ -> Result.top q
+
+
+  (* Thread transfer functions. *)
+
+  let threadenter ctx lval f args =
+    let st = ctx.local in
+    match Cilfacade.find_varinfo_fundec f with
+    | fd ->
+      (* TODO: HACK: Simulate enter_multithreaded for first entering thread to publish global inits before analyzing thread.
+         Otherwise thread is analyzed with no global inits, reading globals gives bot, which turns into top, which might get published...
+         sync `Thread doesn't help us here, it's not specific to entering multithreaded mode.
+         EnterMultithreaded events only execute after threadenter and threadspawn. *)
+      if not (ThreadFlag.is_multi (Analyses.ask_of_ctx ctx)) then
+        ignore (Priv.enter_multithreaded (Analyses.ask_of_ctx ctx) ctx.global ctx.sideg st);
+      let st' = Priv.threadenter (Analyses.ask_of_ctx ctx) ctx.global st in
+      let arg_vars =
+        fd.sformals
+        |> List.filter AD.varinfo_tracked
+        |> List.map RV.arg
+      in
+      let new_apr = AD.add_vars st'.apr arg_vars in
+      [{st' with apr = new_apr}]
+    | exception Not_found ->
+      (* Unknown functions *)
+      (* TODO: do something like base? *)
+      failwith "apron.threadenter: unknown function"
+
+  let threadspawn ctx lval f args fctx =
+    ctx.local
+
+  let event ctx e aprx =
+    let st = ctx.local in
+    match e with
+    | Events.Lock addr when ThreadFlag.is_multi (Analyses.ask_of_ctx ctx) -> (* TODO: is this condition sound? *)
+      Priv.lock (Analyses.ask_of_ctx aprx) aprx.global st addr
+    | Events.Unlock addr when ThreadFlag.is_multi (Analyses.ask_of_ctx ctx) -> (* TODO: is this condition sound? *)
+      Priv.unlock (Analyses.ask_of_ctx aprx) aprx.global aprx.sideg st addr
+    (* No need to handle escape because escaped variables are always referenced but this analysis only considers unreferenced variables. *)
+    | Events.EnterMultiThreaded ->
+      Priv.enter_multithreaded (Analyses.ask_of_ctx aprx) aprx.global aprx.sideg st
+    | _ ->
+      st
+
+  let sync ctx reason =
+    (* After the solver is finished, store the results (for later comparison) *)
+    if !GU.postsolving then begin
+      let old_value = PCU.RH.find_default results ctx.node (AD.bot ()) in
+      let new_value = AD.join old_value ctx.local.apr in
+      PCU.RH.replace results ctx.node new_value;
+    end;
+    Priv.sync (Analyses.ask_of_ctx ctx) ctx.global ctx.sideg ctx.local (reason :> [`Normal | `Join | `Return | `Init | `Thread])
+end
```

Comment:
Usually, we like to have a newline to end each file

Reply:
Should be resolved for all files